pkg/autoscaler/autoscaler.go

```diff
@@ -291,6 +296,138 @@ var _ = Describe("Autoscaler should", framework.LabelAutoscaler, Serial, func()
 		})
 
+		// Machines required for test: 1
+		// Reason: This tests checks that autoscaler is able to scale from zero when a workload requires specific architecture in the node affinity fields.
+		// Moreover, this test gives a better signal when multiple architectures are available in the cluster or the cluster is not amd64,
+		// and the workload requires to be scheduled on an architecture different from amd64.
+		It("It scales from/to zero a machine set with the architecture requested by the workload", func() {
+			// Only run in platforms which support arch-aware autoscaling from/to zero.
+			clusterInfra, err := framework.GetInfrastructure(ctx, client)
+			Expect(err).NotTo(HaveOccurred(), "Failed to get cluster infrastructure object")
+
+			platform := clusterInfra.Status.PlatformStatus.Type
+			switch platform {
+			case configv1.AWSPlatformType, configv1.AzurePlatformType:
+				klog.Infof("Platform is %v", platform)
+			default:
+				Skip(fmt.Sprintf("Platform %v does not support arch-aware autoscaling from/to zero, skipping.", platform))
+			}
+
+			By("Creating a new MachineSet with 0 replicas for each machineset of a different architecture found in the cluster")
+			machineSetParamsList := framework.BuildPerArchMachineSetParamsSet(ctx, client, 0)
+			targetedNodeLabel := fmt.Sprintf("%v-scale-from-zero", autoscalerWorkerNodeRoleLabel)
+			workloadArch := framework.Amd64
+			expectedReplicas := int32(1)
+			var expectedScaledMachineSet *machinev1.MachineSet
+			var machineSets = make([]*machinev1.MachineSet, 0)
+
+			for _, machineSetParams := range machineSetParamsList {
+				machineSetParams.Labels[targetedNodeLabel] = ""
+				By(fmt.Sprintf("Deploying the MachineSet with 0 replicas (architecture: %s)",
+					machineSetParams.Labels[framework.ArchLabel]))
+				machineSet, err := framework.CreateMachineSet(client, machineSetParams)
+				Expect(err).ToNot(HaveOccurred(), "Failed to create MachineSet with 0 replicas")
+				machineSets = append(machineSets, machineSet)
+				cleanupObjects[machineSet.GetName()] = machineSet
+
+				framework.WaitForMachineSet(ctx, client, machineSet.GetName())
+
+				By(fmt.Sprintf("Creating a MachineAutoscaler backed by MachineSet %s/%s - min:%v, max:%v",
+					machineSet.GetNamespace(), machineSet.GetName(), 0, expectedReplicas))
+				asr := machineAutoscalerResource(machineSet, 0, expectedReplicas)
+				Expect(client.Create(ctx, asr)).Should(Succeed(), fmt.Sprintf("Failed to create MachineAutoscaler with min 0/max %v replicas", expectedReplicas))
+				cleanupObjects[asr.GetName()] = asr
+				arch := machineSetParams.Labels[framework.ArchLabel]
+				if arch != framework.Amd64 {
+					workloadArch = arch
+					expectedScaledMachineSet = machineSet
+				}
+			}
+			if workloadArch == "" {
+				// If workloadArch is empty, it means that exactly one MachineSet is in the list, and it is amd64.
```

Comment:
will it be empty though? it seems like we initialize it on L319

Reply:
you're right. I fixed it by moving this condition within the for loop a few lines before and by initializing the workloadArch to the empty string so that the `expectedScaledMachineSet` pointer is set accordingly.

Reply:
You're right. I moved the condition a few lines before, within the for loop, and I initialized the var to the empty string. This way, we ensure the assignment of the `expectedScaledMachineSet` is done at least once.